validators.go

```diff
@@ -208,3 +214,386 @@ func ValidEmail() Email {
+			return IPv4CIDR
+		} else {
+			return IPv4
+		}
+	case has_dot == -1 && has_colon > -1 && l >= 6 && l <= IPv6CIDR:
+		if isWithCIDR(str, l) == true {
+			return IPv6CIDR
+		} else {
+			return IPv6
+		}
+
+	case has_dot > -1 && has_colon > -1 && l >= 14 && l <= IPv4MappedIPv6:
+		if isWithCIDR(str, l) == true {
+			return IPv4MappedIPv6CIDR
+		} else {
+			return IPv4MappedIPv6
+		}
+	}
+
+	return None
+}
+
+func (i IPAddr) IsSatisfied(obj interface{}) bool {
+
+	if str, ok := obj.(string); ok {
+
+		l := len(str)
+		ret := getIPType(str, l)
+
+		for _, ck := range i.vaildtypes {
+
+			if ret != None && (ck == ret || ck == IPAny) {
+
+				switch ret {
+				case IPv4, IPv6, IPv4MappedIPv6:
+					ip := net.ParseIP(str)
+
+					if ip != nil {
+						return true
+					}
+
+				case IPv4CIDR, IPv6CIDR, IPv4MappedIPv6CIDR:
+					_, _, err := net.ParseCIDR(str)
+					if err == nil {
+						return true
+					}
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func (i IPAddr) DefaultMessage() string {
+	return fmt.Sprintln("Must be a vaild IP address")
+}
+
+// Requires a MAC Address string to be exactly
+type MacAddr struct{}
+
+func ValidMacAddr() MacAddr {
+
+	return MacAddr{}
+}
+
+func (m MacAddr) IsSatisfied(obj interface{}) bool {
+
+	if str, ok := obj.(string); ok {
+		if _, err := net.ParseMAC(str); err == nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m MacAddr) DefaultMessage() string {
+	return fmt.Sprintln("Must be a vaild MAC address")
+}
+
+var domainPattern = regexp.MustCompile(`^(([a-zA-Z0-9-\p{L}]{1,63}\.)?(xn--)?[a-zA-Z0-9\p{L}]+(-[a-zA-Z0-9\p{L}]+)*\.)+[a-zA-Z\p{L}]{2,63}$`)
+
+// Requires a Domain string to be exactly
+type Domain struct {
+	Regexp *regexp.Regexp
+}
+
+func ValidDomain() Domain {
+	return Domain{domainPattern}
+}
+
+func (d Domain) IsSatisfied(obj interface{}) bool {
+
+	if str, ok := obj.(string); ok {
+
+		l := len(str)
+		//can't exceed 253 chars.
+		if l > 253 {
+			return false
+		}
+
+		//first and last char must be alphanumeric
+		if str[l-1] == 46 || str[0] == 46 {
+			return false
+		}
+
+		return domainPattern.MatchString(str)
+	}
+
+	return false
+}
+
+func (d Domain) DefaultMessage() string {
+	return fmt.Sprintln("Must be a vaild domain address")
+}
+
+var urlPattern = regexp.MustCompile(`^((((https?|ftps?|gopher|telnet|nntp)://)|(mailto:|news:))(%[0-9A-Fa-f]{2}|[-()_.!~*';/?:@#&=+$,A-Za-z0-9\p{L}])+)([).!';/?:,][[:blank:]])?$`)
+
+type URL struct {
+	Domain
+}
+
+func ValidURL() URL {
+	return URL{Domain: ValidDomain()}
+}
+
+func (u URL) IsSatisfied(obj interface{}) bool {
+
+	if str, ok := obj.(string); ok {
+
+		// TODO : Required lot of testing
+		return urlPattern.MatchString(str)
+	}
+
+	return false
+}
+
+func (u URL) DefaultMessage() string {
+	return fmt.Sprintln("Must be a vaild URL address")
+}
+
+/*
+NORMAL BenchmarkRegex-8   	2000000000	         0.24 ns/op
+STRICT BenchmarkLoop-8    	2000000000	         0.01 ns/op
+*/
+const (
+	NORMAL = 0
+	STRICT = 4
+)
+
```

Comment:
The difference between them as follow

STRICT : Not allow Invisible characters (control characters without tab, space, newline)
NORMAL : Allow Invisible characters

Welcome back~ @pedromorgan